perception/lidar_transfusion/include/lidar_transfusion/transfusion_trt.hpp

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIDAR_TRANSFUSION__TRANSFUSION_TRT_HPP_
+#define LIDAR_TRANSFUSION__TRANSFUSION_TRT_HPP_
+
+#include "lidar_transfusion/cuda_utils.hpp"
+#include "lidar_transfusion/network/network_trt.hpp"
+#include "lidar_transfusion/postprocess/postprocess_kernel.hpp"
+#include "lidar_transfusion/preprocess/pointcloud_densification.hpp"
+#include "lidar_transfusion/preprocess/preprocess_kernel.hpp"
+#include "lidar_transfusion/preprocess/voxel_generator.hpp"
+#include "lidar_transfusion/utils.hpp"
+#include "lidar_transfusion/visibility_control.hpp"
+
+#include <tier4_autoware_utils/system/stop_watch.hpp>
+
+#include <sensor_msgs/msg/point_cloud2.hpp>
+
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace lidar_transfusion
+{
+
+class NetworkParam
+{
+public:
+  NetworkParam(std::string onnx_path, std::string engine_path, std::string trt_precision)
+  : onnx_path_(std::move(onnx_path)),
+    engine_path_(std::move(engine_path)),
+    trt_precision_(std::move(trt_precision))
+  {
+  }
+
+  std::string onnx_path() const { return onnx_path_; }
+  std::string engine_path() const { return engine_path_; }
+  std::string trt_precision() const { return trt_precision_; }
+
+private:
+  std::string onnx_path_;
+  std::string engine_path_;
+  std::string trt_precision_;
+};
+
+class LIDAR_TRANSFUSION_PUBLIC TransfusionTRT
+{
+public:
+  explicit TransfusionTRT(
+    const NetworkParam & network_param, const DensificationParam & densification_param,
+    const TransfusionConfig & config);
+  virtual ~TransfusionTRT();
+
+  bool detect(
+    const sensor_msgs::msg::PointCloud2 & msg, const tf2_ros::Buffer & tf_buffer,
+    std::vector<Box3D> & det_boxes3d, std::unordered_map<std::string, double> & proc_timing);
+
+protected:
+  void initPtr();
+
+  bool preprocess(const sensor_msgs::msg::PointCloud2 & msg, const tf2_ros::Buffer & tf_buffer);
+
+  bool inference();
+
+  bool postprocess(std::vector<Box3D> & det_boxes3d);
+  bool postprocessCPU(std::vector<Box3D> & det_boxes3d);
+
+  std::unique_ptr<NetworkTRT> network_trt_ptr_{nullptr};
+  std::unique_ptr<VoxelGenerator> vg_ptr_{nullptr};
+  std::unique_ptr<tier4_autoware_utils::StopWatch<std::chrono::milliseconds>> stop_watch_ptr_;
+  std::shared_ptr<PreprocessCuda> pre_ptr_;
```

Comment:
any reason these should not be unique ptr?

Reply:
also uniform initialization to nullptr

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6890/commits/48200d207d459c01f9fb7dc38b33b355bc1437f0